File: matiasmasso/Mat.NET/DT2/DTOPostComment.vb
Public Class DTOPostComment
    Inherits DTOBaseGuid
    Property Parent As Guid
    Property ParentSource As ParentSources
    Property User As DTOUser
    Property Lang As DTOLang
    Property Fch As Date
    Property Text As String
    Property FchApproved As Date
    Property FchDeleted As Date
    Property Answering As DTOPostComment
    Property Answers As New List(Of DTOPostComment)

    Private _ParentTitle As DTOLangText

    Property ErrorCode As ErrorCodes

    Public Enum StatusEnum
        NotSet
        Pendent
        Aprobat
        Eliminat
    End Enum

    Public Enum ParentSources
        NotSet
        Noticia
        Blog
    End Enum

    Public Enum ErrorCodes
        None
        NotLogged
    End Enum

    Public Sub New()
        MyBase.New
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function NewFromParent(oParent As Guid) As DTOPostComment
        Dim retval As New DTOPostComment()
        retval.Parent = oParent
        Return retval
    End Function

    Shared Function NewFromParentSource(oGuid As Guid) As DTOPostComment
        Dim retval As New DTOPostComment()
        retval.Parent = oGuid
        retval.Fch = Now
        Return retval
    End Function

    Shared Function FromQuestion(oAnsweringComment As DTOPostComment, sText As String) As DTOPostComment
        Dim retval As New DTOPostComment
        With retval
            .Answering = oAnsweringComment
            .Parent = oAnsweringComment.Parent
            .ParentSource = oAnsweringComment.ParentSource
            .User = DTOUser.wellknown(DTOUser.wellknowns.matias)
            .FchApproved = Now
            .Fch = Now
            .Text = sText
        End With
        Return retval
    End Function

    Public ReadOnly Property Status() As StatusEnum
        Get
            Dim retval As StatusEnum = StatusEnum.NotSet
            If _FchDeleted = Nothing Then
                If _FchApproved = Nothing Then
                    retval = StatusEnum.Pendent
                Else
                    retval = StatusEnum.Aprobat
                End If
            Else
                retval = StatusEnum.Eliminat
            End If
            Return retval
        End Get
    End Property

    Public ReadOnly Property NickName As String
        Get
            Dim retval As String = ""
            If _User IsNot Nothing Then
                retval = _User.nickName
            End If
            Return retval
        End Get
    End Property


    Public Property ParentTitle As DTOLangText
        Get
            If _ParentTitle Is Nothing Then
                Dim oNoticia As New DTONoticia(_Parent)
                _ParentTitle = oNoticia.title
            End If
            Return _ParentTitle
        End Get
        Set(value As DTOLangText)
            _ParentTitle = value
        End Set
    End Property


    Shared Function HtmlText(oComment As DTOPostComment) As String
        Dim src As String = oComment.Text
        Dim retval As String = src.Replace(vbCrLf, "<br/>")
        Return retval
    End Function

    Shared Function Noticia(oPostComment As DTOPostComment) As DTONoticia
        Dim retval As New DTONoticia(oPostComment.Parent)
        Return retval
    End Function

    Shared Function UserNickname(oPostComment As DTOPostComment) As String
        Dim oUser As DTOUser = oPostComment.User
        Dim retval As String = oUser.nickName
        If retval = "" Then retval = oUser.nom
        If retval = "" Then retval = oUser.emailAddress.Substring(0, oUser.emailAddress.IndexOf("@") - 1)
        Return retval
    End Function

    Shared Function Tree(items As List(Of DTOPostComment)) As List(Of DTOPostComment)
        'construeix arbre de comentaris a partir de una llista plana

        Dim retval As New List(Of DTOPostComment)
        For Each oItem As DTOPostComment In items
            If oItem.Answering Is Nothing Then
                'afageix-lo com a comentari arrel de conversa
                retval.Insert(0, oItem)
            Else
                'assigna les respostes als seus comentaris respectius
                For Each oTarget As DTOPostComment In items
                    If oTarget.Equals(oItem.Answering) Then
                        oTarget.Answers.Add(oItem)
                    End If
                Next
            End If
        Next

        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOPostComment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DTO
{

    public class DTOPostComment : DTOBaseGuid
    {
        public Guid Parent { get; set; }
        public ParentSources ParentSource { get; set; }
        public DTOUser User { get; set; }
        public DTOLang Lang { get; set; }
        public DateTime Fch { get; set; }
        public string Text { get; set; }
        public DateTime FchApproved { get; set; }
        public DateTime FchDeleted { get; set; }
        public DTOPostComment Answering { get; set; }
        public DTOPostComment AnswerRoot { get; set; }
        public List<DTOPostComment> Answers { get; set; } = new List<DTOPostComment>();

        private DTOLangText _ParentTitle;

        public int CommentsCount { get; set; }
        public int Idx { get; set; }


        public ErrorCodes ErrorCode { get; set; }

        public enum StatusEnum
        {
            NotSet,
            Pendent,
            Aprobat,
            Eliminat
        }

        public enum ParentSources
        {
            NotSet,
            Noticia,
            News,
            Blog,
            Shop4moms
        }

        public enum ErrorCodes
        {
            None,
            NotLogged
        }

        public DTOPostComment() : base()
        {
        }

        public DTOPostComment(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOPostComment NewFromParent(Guid oParent)
        {
            DTOPostComment retval = new DTOPostComment();
            retval.Parent = oParent;
            return retval;
        }

        public static DTOPostComment NewFromParentSource(Guid oGuid)
        {
            DTOPostComment retval = new DTOPostComment();
            retval.Parent = oGuid;
            retval.Fch = DTO.GlobalVariables.Now();
            return retval;
        }

        public static DTOPostComment FromQuestion(DTOPostComment oAnsweringComment, string sText)
        {
            DTOPostComment retval = new DTOPostComment();
            {
                var withBlock = retval;
                withBlock.Answering = oAnsweringComment;
                withBlock.Parent = oAnsweringComment.Parent;
                withBlock.ParentSource = oAnsweringComment.ParentSource;
                withBlock.User = DTOUser.Wellknown(DTOUser.Wellknowns.matias);
                withBlock.FchApproved = DTO.GlobalVariables.Now();
                withBlock.Fch = DTO.GlobalVariables.Now();
                withBlock.Text = sText;
            }
            return retval;
        }

        public string Url(DTOLang lang, bool absoluteUrl = false)
        {
            DTOWebDomain domain = DTOWebDomain.Factory(lang, absoluteUrl);
            string retval = domain.Url("PostComment", base.Guid.ToString());
            return retval;
        }

        public StatusEnum Status
        {
            get
            {
                StatusEnum retval = StatusEnum.NotSet;
                if (FchDeleted == default(DateTime))
                {
                    if (FchApproved == default(DateTime))
                        retval = StatusEnum.Pendent;
                    else
                        retval = StatusEnum.Aprobat;
                }
                else
                    retval = StatusEnum.Eliminat;
                return retval;
            }
        }

        public string NickName
        {
            get
            {
                string retval = "";
                if (User != null)
                    retval = User.NickName;
                return retval;
            }
        }


        public DTOLangText ParentTitle
        {
            get
            {
                if (_ParentTitle == null)
                {
                    DTONoticia oNoticia = new DTONoticia(Parent);
                    _ParentTitle = oNoticia.Title;
                }
                return _ParentTitle;
            }
            set
            {
                _ParentTitle = value;
            }
        }


        public string HtmlText()
        {
            return HtmlText(this);
        }
        public static string HtmlText(DTOPostComment oComment)
        {
            string src = oComment.Text;
            string retval = System.Text.RegularExpressions.Regex.Replace(src, @"\r\n?|\n", "<br />"); // src.Replace(Constants.vbCrLf, "<br/>");
            return retval;
        }

        public static DTONoticia Noticia(DTOPostComment oPostComment)
        {
            DTONoticia retval = new DTONoticia(oPostComment.Parent);
            return retval;
        }

        public static string UserNickname(DTOPostComment oPostComment)
        {
            DTOUser oUser = oPostComment.User;
            string retval = oUser.NickName;
            if (retval == "")
                retval = oUser.Nom;
            if (retval == "")
                retval = oUser.EmailAddress.Substring(0, oUser.EmailAddress.IndexOf("@") - 1);
            return retval;
        }



        public string ShortenedText(int len = 99)
        {
            string retval = this.Text;
            if (retval.Length > len)
            {
                if (this.Text.IndexOf(" ", 99) > 0)
                {
                    retval = retval.Substring(0, this.Text.IndexOf(" ", 99));
                    if (retval.Length < this.Text.Length)
                        retval += "...";
                }
            }
            return retval;
        }

        public static int RecursiveCount(List<DTOPostComment> comments)
        {
            int retval = comments.Count();
            foreach (DTOPostComment item in comments)
            {
                retval += RecursiveCount(item.Answers);
            }
            return retval;
        }

        //public SyndicationItem Rss(DTOWebDomain domain)
        //{
        //    var lang = domain.DefaultLang();
        //    var retval = new SyndicationItem()
        //    {
        //        Id = this.Guid.ToString(),
        //        Title = SyndicationContent.CreatePlaintextContent(this.ParentTitle.Tradueix(lang)),
        //        Content = SyndicationContent.CreateHtmlContent(this.Text),
        //        PublishDate = this.Fch,
        //        LastUpdatedTime = this.Fch
        //    };
        //    retval.Links.Add(SyndicationLink.CreateAlternateLink(new Uri(Url(domain.DefaultLang(), true))));
        //    return retval;
        //}

        //------------------------------------------------------------------------

        public class TreeModel
        {
            public DTOBaseGuid Target { get; set; }
            public DTOPostComment.ParentSources TargetSrc { get; set; }
            public List<Item> Items { get; set; }
            public int Take { get; set; }
            public int From { get; set; } //0 based
            public int RootItemsCount { get; set; }

            public TreeModel() : base()
            {
                this.Take = 15; //fetch comments from 15 to 15
                this.From = 0; //zero based
                this.Items = new List<Item>();
            }

            public int ItemsLeftCount()
            {
                return this.RootItemsCount - this.Items.Count - this.From;
            }

            public int NextCount()
            {
                //per el link ver {NextCount] más...
                int retval = Take < ItemsLeftCount() ? Take : ItemsLeftCount();
                return retval;
            }

            public bool MoreItems()
            {
                return (NextCount() > 0);
            }

            public class Item
            {
                public Guid Guid { get; set; }
                public int Idx { get; set; }
                public int Siblings { get; set; }
                public DateTime Fch { get; set; }
                public Author Author { get; set; }
                public string Text { get; set; }
                public Item Answering { get; set; }
                public Guid AnswerRootGuid { get; set; }
                public List<Item> Items { get; set; }

                public Item()
                {
                    this.Items = new List<Item>();
                }

                public static Item Factory(DTOPostComment comment)
                {
                    Item retval = new Item();
                    retval.Guid = comment.Guid;
                    if (retval.Guid != Guid.Empty)
                    {
                        retval.Author = new Author();
                        retval.Author.Guid = comment.User.Guid;
                        retval.Author.Nom = comment.User.NicknameForComments();
                        retval.Fch = comment.Fch;
                        retval.Text = comment.Text;
                        if (comment.Answering != null)
                        {
                            retval.Answering = Item.Factory(comment.Answering);
                        }
                        if (comment.AnswerRoot != null)
                        {
                            retval.AnswerRootGuid = comment.AnswerRoot.Guid;
                        }
                    }
                    return retval;
                }


                public string ShortenedText(int len = 99)
                {
                    string retval = "";
                    if (!string.IsNullOrEmpty(this.Text))
                    {
                        retval = this.Text;
                        if (retval.Length > len)
                        {
                            if (this.Text.IndexOf(" ", 99) > 0)
                            {
                                retval = retval.Substring(0, this.Text.IndexOf(" ", 99));
                                if (retval.Length < this.Text.Length)
                                    retval += "...";
                            }
                        }
                    }
                    return retval;
                }

                public string Caption(DTOLang lang)
                {
                    StringBuilder sb = new StringBuilder();
                    if (this.Author != null)
                    {
                        sb.AppendFormat("{0:dd/MM/yy} {0:HH:mm} {1}", this.Fch, this.Author.Nom);
                        if (this.Answering != null)
                        {
                            sb.AppendFormat(" {0} {1}", lang.Tradueix("en respuesta a", "en resposta a", "answering"), this.Answering.Author.Nom);
                        }
                    }
                    return sb.ToString();
                }
            }

            public class Author
            {
                public Guid Guid { get; set; }
                public string Nom { get; set; }

                public Author() : base()
                {

                }
            }

        }
    }


}